controller/lib/hal/hal_stm32.cpp

```diff
@@ -0,0 +1,856 @@
+{
+   TimerRegs *tmr = (TimerRegs *)TIMER6_BASE;
+   tmr->status = 0;
+   msCount++;
+}
+
+uint32_t HalApi::millis(){
+   return msCount;
+}
+
+void HalApi::delay(uint32_t ms){
+   uint32_t start = msCount;
+   while( 1 )
+   {
+      uint32_t dt = msCount - start;
+      if( dt >= ms )
+         return;
+   }
+}
+
+/******************************************************************
+ * A/D inputs.
+ *
+ * The following pins are used as analog inputs on the rev-1 PCB 
+ *
+ * PA0 (ADC1_IN5)  - vin
+ * PA1 (ADC1_IN6)  - pressure
+ * PA4 (ADC1_IN9)  - inhale flow
+ * PB0 (ADC1_IN15) - exhale flow
+ *
+ *****************************************************************/
+static void InitADC( void ){
+   // Enable the clock to the A/D converter
+   EnableClock( ADC_BASE );
+
+   // Configure the 4 pins used as analog inputs 
+   GPIO_PinMode( GPIO_A_BASE, 0, GPIO_MODE_ANALOG );
+   GPIO_PinMode( GPIO_A_BASE, 1, GPIO_MODE_ANALOG );
+   GPIO_PinMode( GPIO_A_BASE, 4, GPIO_MODE_ANALOG );
+   GPIO_PinMode( GPIO_B_BASE, 0, GPIO_MODE_ANALOG );
+
+   // Perform a power-up and calibration sequence on 
+   // the A/D converter
+   ADC_Regs *adc = (ADC_Regs *)ADC_BASE;
+
+   // Exit deep power down mode and turn on the
+   // internal voltage regulator.
+   adc->adc[0].ctrl   = 0x10000000;
+
+   // Wait for the startup time specified in the datasheet
+   // for the voltage regulator to become ready.
+   // The time in the datasheet is 20 microseconds, but
+   // I'll wait for 30 just to be extra conservative
+   BusyWait( 30 );
+
+   // Calibrate the A/D for single ended channels
+   adc->adc[0].ctrl |= 0x80000000;
+
+   // Wait until the CAL bit is cleared meaning
+   // calibration is finished
+   while( adc->adc[0].ctrl & 0x80000000 ){}
+
+   // Clear all the status bits
+   adc->adc[0].stat = 0x3FF;
+
+   // Enable the A/D
+   adc->adc[0].ctrl |= 0x00000001;
+
+   // Wait for the ADRDY status bit to be set
+   while( !(adc->adc[0].stat & 0x00000001) ){}
+
+   // Configure the A/D as 12-bit resolution
+   adc->adc[0].cfg[0] = 0x00000000;
+
+   // Set sample time. I'm using 92.5 A/D clocks (a little over 1us)
+   // to sample.  We'll need to do a bit of testing to see what the
+   // best value is for this
+   adc->adc[0].samp[0] = 5;
+}
+
+// Read a single A/D channel.
+// This just does the conversion immediately when called.
+//
+// It would be better to start a sequence of conversions at the beginning of the loop
+// but the HAL interface currently isn't set up to run that way.  That's an improvement
+// that can be made later
+int HalApi::analogRead(AnalogPin pin){
+
+   int channel = 0;
+   switch( pin )
+   {
+      case AnalogPin::PATIENT_PRESSURE:
+         channel = 6;
+         break;
+
+      case AnalogPin::INFLOW_PRESSURE_DIFF:
+         channel = 9;
+         break;
+
+      case AnalogPin::OUTFLOW_PRESSURE_DIFF:
+         channel = 15;
+         break;
+
+      default:
+         return 0;
+   }
+
+   ADC_Regs *adc = (ADC_Regs *)ADC_BASE;
```

Comment:
I'm not really sure how to do that other then creating some nasty pre-defined symbols in the linker file at specific addresses.  This seems like a much cleaner way to handle it to me.
Maybe I'm just not understanding what you're asking.  How (other then a cast) would I create a pointer to a structure at a specific memory location?

Reply:
Oh, I just meant, instead of `#define ADC_BASE 0xdeadbeef`, could we do `const inline ADC_Regs* ADC_BASE = reinterpret_cast<ADC_Regs*>(0xdeadbeef);`?  (Can't be constexpr because reinterpret_cast is not allowed in constexpr.)

Reply:
OK, I see what you're saying, but does this really help at all?  We're basically just moving the cast from the cpp file to the header.  I'd prefer to leave it like it is, but will change if you feel strongly about it.

Reply:
struct ADC_Regs is effectively a singleton, right?  There is only one valid instance, namely the one at address ADC_BASE?

Putting the cast into the header file right next to the definition of the struct is how I would tell readers the above -- that ADC_Regs is a singleton whose instance must be at address 0xdeadbeef.  Then anyone else who wants to access ADC_Regs need not concern themselves with the implementation detail of taking a pointer and reinterpret_cast'ing it to the correct type.